staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
+
+	signalFn := s.overrideTerminationEventSignalFunc
+	if signalFn == nil {
+		signalFn = func() {
+			// wait until we receive a KILL signal
+			<-stopCh
+			events.WithName(ShutdownInitiated).Signal()
+
+			// fire AfterShutdownDelayDuration event when ShutdownDelayDuration elapses
+			<-time.After(s.ShutdownDelayDuration)
+			events.WithName(AfterShutdownDelayDuration).Signal()
+
+			// wait until all existing request(s) in flight that we account for are drained.
+			s.HandlerChainWaitGroup.Wait()
+			events.WithName(InFlightRequestsDrained).Signal()
+		}
+	}
+	go signalFn()
 
 	go func() {
-		defer close(delayedStopCh)
-
-		<-stopCh
-
 		// As soon as shutdown is initiated, /readyz should start returning failure.
-		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
-		// and stop sending traffic to this server.
-		close(s.readinessStopCh)
+		// This gives the load balancer a window defined by ShutdownDelayDuration to
+		// detect that /readyz is red and stop sending traffic to this server.
+		defer close(s.readinessStopCh)
 
-		time.Sleep(s.ShutdownDelayDuration)
+		<-events.WithName(ShutdownInitiated).Signaled()
```

Comment:
mixing `the new abstraction` and `regular chan` makes it a bit hard to read

Reply:
yes, i agree. I simplified a bit, let me know if it reads better. I guess this is a tradeoff, we need the abstraction for a flake free test, on the other hand, we lose nice idiomatic inline channel and go func.